File: packages/engine/src/core/index-relevance.ts
import type { JudgementRecordedPayload } from '@sofar/schema'
import { passOverRecord } from './index-pass'
import { INDEX_SCHEMA_VERSION, readIndexFile, writeIndexFile } from './index-store'
import type { IndexedEvent } from './index-tail'

/**
 * Stored relevance, keyed for lookup (typed-judge 5.1, the D10 contract).
 *
 * A relevance judgement is made at write time (the write-back pass,
 * core/relevance-judge.ts) or server-side, and read where no judge may run:
 * the SessionStart digest (5.3) and memory-lead B1's read-time surfacing,
 * both on hooks (typed-judge D1). This tier is what they read: one row per
 * (about, subject), the latest judgement winning, from judgement_recorded
 * events with question `relevance`. Derived and disposable like every tier
 * (record-index D1).
 *
 * Two laws from D10 live in the reader, not the tier:
 *
 * - RETIRED STAYS RETIRED. The reader never returns a superseded or
 *   until-retired decision, whatever its stored p. It does not re-derive
 *   retirement: that is the fold's supersession logic, and copying it into a
 *   reducer is the drift the index's faithful-not-better law forbids. So the
 *   caller passes the retired handles, from the fold it already holds
 *   (SessionStart) or from its own structure (B1). The parameter is required.
 * - RANK AND ADD, NEVER REMOVE. `rankByRelevance` keeps every deterministic
 *   candidate (guard, derived scope, lexical link) and only orders it, and
 *   adds a candidate the judge found only at p >= RELEVANCE_CARRY. The drop
 *   at p <= relevance_drop can therefore touch only what the judge would have
 *   added.
 */

/**
 * A candidate the judge alone found is added at or above this: THRESHOLDS.
 * relevance_carry (typed-judge 1.2, jev-1.13.0). Restated rather than imported,
 * because hooks read this module and must never reach core/judge.ts (D1's
 * import pin). A test pins the two equal.
 */
export const RELEVANCE_CARRY = 0.8

const RELEVANCE_FILE = 'relevance.json'
const RELEVANCE_META = 'meta-relevance.json'

/** One stored answer, as the tier keeps it. */
export interface RelevanceRow {
  /** Qualified handle, event id or task id: bare `D<n>` is qualified with the writing initiative. */
  subject: string
  /** P(relevant). */
  p: number
  model: string
  /** The judgement_recorded event id: later ids win. */
  id: string
}

interface SlugRelevanceState {
  /** about → subject → row. */
  rows: Record<string, Record<string, RelevanceRow>>
}

interface RelevanceDisk {
  version: number
  initiatives: Record<string, SlugRelevanceState>
}

function isRelevanceDisk(v: unknown): v is RelevanceDisk {
  if (typeof v !== 'object' || v === null) return false
  const r = v as Record<string, unknown>
  return r.version === INDEX_SCHEMA_VERSION && typeof r.initiatives === 'object' && r.initiatives !== null
}

/** A bare `D12` names the envelope's own initiative (the citation grammar); qualify it so rows are comparable repo-wide. */
export function qualifySubject(subject: string, slug: string): string {
  return /^D\d+$/.test(subject) ? `${slug} ${subject}` : subject
}

function apply(state: SlugRelevanceState, event: IndexedEvent, slug: string): void {
  if (event.type !== 'judgement_recorded') return
  const p = event.payload as unknown as JudgementRecordedPayload
  if (p.question !== 'relevance' || typeof p.about !== 'string' || p.answer?.type !== 'noul') return
  if (typeof p.subject !== 'string' || typeof p.model !== 'string') return
  const subject = qualifySubject(p.subject, slug)
  const rows = state.rows[p.about] ?? {}
  const prior = rows[subject]
  if (prior === undefined || event.id > prior.id) rows[subject] = { subject, p: p.answer.noul, model: p.model, id: event.id }
  state.rows[p.about] = rows
}

const empty = (): SlugRelevanceState => ({ rows: {} })
const clone = (s: SlugRelevanceState): SlugRelevanceState => ({
  rows: Object.fromEntries(Object.entries(s.rows).map(([about, rows]) => [about, { ...rows }])),
})

/** Per-initiative rows: a `task:` row is about a task of the initiative that wrote it. */
export type RelevanceIndex = Record<string, SlugRelevanceState>

/** Bring the tier up to date. Cost is O(events since its own cursor). */
export function refreshRelevance(sofarDir: string): RelevanceIndex {
  const prior = readIndexFile<RelevanceDisk>(sofarDir, RELEVANCE_FILE, isRelevanceDisk)
  const { states, changed } = passOverRecord<SlugRelevanceState>(sofarDir, RELEVANCE_META, prior === null ? null : prior.initiatives, {
    empty,
    clone,
    apply,
  })
  if (changed) writeIndexFile(sofarDir, RELEVANCE_FILE, { version: INDEX_SCHEMA_VERSION, initiatives: states })
  return states
}

/** Read the tier without refreshing; null when nothing usable is on disk. */
export function readRelevance(sofarDir: string): RelevanceIndex | null {
  const disk = readIndexFile<RelevanceDisk>(sofarDir, RELEVANCE_FILE, isRelevanceDisk)
  return disk === null ? null : disk.initiatives
}

/**
 * The stored rows about `about`, strongest first, never a retired subject.
 * `task:` rows belong to one initiative, so `initiative` names it; `file:` rows
 * are repo-wide and every initiative's count. `retired` holds qualified
 * handles (`<slug> D<n>`) and is required: see the header.
 */
export function relevance(
  index: RelevanceIndex,
  query: { about: string; initiative?: string; retired: ReadonlySet<string> },
): RelevanceRow[] {
  const slugs = query.about.startsWith('task:') ? (query.initiative !== undefined ? [query.initiative] : []) : Object.keys(index)
  const best = new Map<string, RelevanceRow>()
  for (const slug of slugs) {
    for (const row of Object.values(index[slug]?.rows[query.about] ?? {})) {
      if (query.retired.has(row.subject)) continue
      const prior = best.get(row.subject)
      if (prior === undefined || row.id > prior.id) best.set(row.subject, row)
    }
  }
  return [...best.values()].sort((a, b) => b.p - a.p || (a.subject < b.subject ? -1 : 1))
}

/**
 * Rank and add (D10). Every deterministic candidate stays, ordered by its
 * stored p, with no row counting as undecided (0.5) and ties keeping the
 * deterministic order. A row naming something that is not a candidate is
 * added only at p >= RELEVANCE_CARRY. Nothing is removed.
 */
export function rankByRelevance(candidates: readonly string[], rows: readonly RelevanceRow[]): string[] {
  const p = new Map(rows.map((r) => [r.subject, r.p]))
  const known = new Set(candidates)
  const added = rows.filter((r) => !known.has(r.subject) && r.p >= RELEVANCE_CARRY).map((r) => r.subject)
  const all = [...candidates, ...added]
  const position = new Map(all.map((s, i) => [s, i]))
  return all.sort((a, b) => (p.get(b) ?? 0.5) - (p.get(a) ?? 0.5) || position.get(a)! - position.get(b)!)
}

File: crates/sofar-core/src/index_relevance.rs
//! Stored relevance, keyed for lookup (`core/index-relevance.ts`, typed-judge
//! 5.1, the D10 contract): one row per (about, subject), the latest
//! `judgement_recorded` with question `relevance` winning. Read where no judge
//! may run — here, the `PostToolUse` notice's reordering within a tier. Two laws
//! live in the reader: a retired subject is never returned (the caller passes
//! the retired handles), and ranking adds or orders, never removes.

use std::collections::HashSet;

use crate::index_pass::{PassResult, SlugReducer, pass_over_record};
use crate::index_tail::IndexedEvent;
use crate::index_tier1::{read_half, write_half};
use crate::json::{Json, Object};
use crate::layout::Layout;
use crate::text::cmp_utf16;

/// `RELEVANCE_CARRY`: a candidate the judge alone found is added at or above this.
pub const RELEVANCE_CARRY: f64 = 0.8;

const RELEVANCE_FILE: &str = "relevance.json";
const RELEVANCE_META: &str = "meta-relevance.json";

/// One stored answer (`RelevanceRow`).
#[derive(Debug, Clone, PartialEq)]
pub struct RelevanceRow {
    pub subject: String,
    pub p: f64,
    pub model: String,
    /// The `judgement_recorded` event id: later ids win.
    pub id: String,
}

/// `SlugRelevanceState`: about → subject → row, each map in insertion order
/// (serialized in JavaScript property order by the JSON layer).
#[derive(Debug, Clone, Default, PartialEq)]
pub struct SlugRelevanceState {
    pub rows: Vec<(String, Vec<(String, RelevanceRow)>)>,
}

impl SlugRelevanceState {
    fn to_json(&self) -> Json {
        let mut rows = Object::with_capacity(self.rows.len());
        for (about, by_subject) in &self.rows {
            let mut o = Object::with_capacity(by_subject.len());
            for (subject, row) in by_subject {
                let mut r = Object::with_capacity(4);
                r.insert("subject", Json::Str(row.subject.clone()));
                r.insert("p", Json::Num(row.p));
                r.insert("model", Json::Str(row.model.clone()));
                r.insert("id", Json::Str(row.id.clone()));
                o.insert(subject.clone(), Json::Obj(r));
            }
            rows.insert(about.clone(), Json::Obj(o));
        }
        let mut o = Object::with_capacity(1);
        o.insert("rows", Json::Obj(rows));
        Json::Obj(o)
    }

    fn from_json(v: &Json) -> Option<Self> {
        let rows = v.as_obj()?.get("rows")?.as_obj()?;
        let mut out = Vec::with_capacity(rows.len());
        for (about, by_subject) in rows.js_ordered() {
            let mut list = Vec::new();
            for (subject, row) in by_subject.as_obj()?.js_ordered() {
                let r = row.as_obj()?;
                list.push((
                    subject.to_owned(),
                    RelevanceRow {
                        subject: r.get("subject")?.as_str()?.to_owned(),
                        p: r.get("p")?.as_f64()?,
                        model: r.get("model")?.as_str()?.to_owned(),
                        id: r.get("id")?.as_str()?.to_owned(),
                    },
                ));
            }
            out.push((about.to_owned(), list));
        }
        Some(Self { rows: out })
    }
}

/// `qualifySubject`: a bare `D12` names the envelope's own initiative.
#[must_use]
pub fn qualify_subject(subject: &str, slug: &str) -> String {
    let bare = subject
        .strip_prefix('D')
        .is_some_and(|d| !d.is_empty() && d.bytes().all(|b| b.is_ascii_digit()));
    if bare {
        format!("{slug} {subject}")
    } else {
        subject.to_owned()
    }
}

struct RelevanceReducer;
impl SlugReducer for RelevanceReducer {
    type State = SlugRelevanceState;
    fn empty(&self) -> SlugRelevanceState {
        SlugRelevanceState::default()
    }
    fn apply(&self, state: &mut SlugRelevanceState, event: &IndexedEvent, slug: &str) {
        if event.event_type != "judgement_recorded" {
            return;
        }
        let p = &event.payload;
        let answer = p.get("answer").and_then(Json::as_obj);
        let (Some("relevance"), Some(about), Some("noul")) = (
            p.get("question").and_then(Json::as_str),
            p.get("about").and_then(Json::as_str),
            answer.and_then(|a| a.get("type")).and_then(Json::as_str),
        ) else {
            return;
        };
        let (Some(subject), Some(model)) = (
            p.get("subject").and_then(Json::as_str),
            p.get("model").and_then(Json::as_str),
        ) else {
            return;
        };
        let subject = qualify_subject(subject, slug);
        let noul = answer
            .and_then(|a| a.get("noul"))
            .and_then(Json::as_f64)
            .unwrap_or(f64::NAN);
        let at = if let Some(i) = state.rows.iter().position(|(a, _)| a == about) {
            i
        } else {
            state.rows.push((about.to_owned(), Vec::new()));
            state.rows.len() - 1
        };
        let rows = &mut state.rows[at].1;
        let row = RelevanceRow {
            subject: subject.clone(),
            p: noul,
            model: model.to_owned(),
            id: event.id.clone(),
        };
        match rows.iter_mut().find(|(s, _)| *s == subject) {
            Some((_, prior)) => {
                if event.id.as_str() > prior.id.as_str() {
                    *prior = row;
                }
            }
            None => rows.push((subject, row)),
        }
    }
}

/// `refreshRelevance`: bring the tier up to date.
#[must_use]
pub fn refresh_relevance(layout: &Layout) -> Vec<(String, SlugRelevanceState)> {
    let prior = read_half(layout, RELEVANCE_FILE, SlugRelevanceState::from_json);
    let PassResult {
        states, changed, ..
    } = pass_over_record(layout, RELEVANCE_META, prior.as_deref(), &RelevanceReducer);
    if changed {
        write_half(layout, RELEVANCE_FILE, &states, SlugRelevanceState::to_json);
    }
    states
}

/// `relevance`: the stored rows about `about`, strongest first, never a
/// retired subject. `task:` rows belong to `initiative`; `file:` rows are
/// repo-wide.
#[must_use]
pub fn relevance<S: std::hash::BuildHasher>(
    index: &[(String, SlugRelevanceState)],
    about: &str,
    initiative: Option<&str>,
    retired: &HashSet<String, S>,
) -> Vec<RelevanceRow> {
    let slugs: Vec<&str> = if about.starts_with("task:") {
        initiative.into_iter().collect()
    } else {
        index.iter().map(|(s, _)| s.as_str()).collect()
    };
    let mut best: Vec<RelevanceRow> = Vec::new();
    for slug in slugs {
        let Some((_, state)) = index.iter().find(|(s, _)| s == slug) else {
            continue;
        };
        let Some((_, rows)) = state.rows.iter().find(|(a, _)| a == about) else {
            continue;
        };
        // Visit order cannot matter: rows dedupe by subject on the id, and
        // the final sort is total.
        for (_, row) in rows {
            if retired.contains(&row.subject) {
                continue;
            }
            match best.iter_mut().find(|b| b.subject == row.subject) {
                Some(prior) => {
                    if row.id > prior.id {
                        *prior = row.clone();
                    }
                }
                None => best.push(row.clone()),
            }
        }
    }
    best.sort_by(|a, b| {
        b.p.partial_cmp(&a.p)
            .filter(|o| o.is_ne())
            .unwrap_or_else(|| {
                if cmp_utf16(&a.subject, &b.subject).is_lt() {
                    std::cmp::Ordering::Less
                } else {
                    std::cmp::Ordering::Greater
                }
            })
    });
    best
}

/// `rankByRelevance` (D10): every candidate stays, ordered by its stored p
/// (no row is 0.5, ties keep the candidates' order); a row naming a
/// non-candidate is added only at p >= `RELEVANCE_CARRY`.
#[must_use]
pub fn rank_by_relevance(candidates: &[String], rows: &[RelevanceRow]) -> Vec<String> {
    let p_of = |s: &str| -> f64 {
        // `new Map(rows.map(...))`: a later row for the same subject wins.
        rows.iter()
            .rev()
            .find(|r| r.subject == s)
            .map_or(0.5, |r| r.p)
    };
    let mut all: Vec<String> = candidates.to_vec();
    for r in rows {
        if !candidates.contains(&r.subject) && r.p >= RELEVANCE_CARRY {
            all.push(r.subject.clone());
        }
    }
    let position = |s: &str, all: &[String]| all.iter().position(|x| x == s).unwrap_or(0);
    let snapshot = all.clone();
    all.sort_by(|a, b| {
        let d = p_of(b) - p_of(a);
        if d != 0.0 && !d.is_nan() {
            return if d < 0.0 {
                std::cmp::Ordering::Less
            } else {
                std::cmp::Ordering::Greater
            };
        }
        position(a, &snapshot).cmp(&position(b, &snapshot))
    });
    all
}
